File: algo_7_bst/main.go
// package main

// type BST struct {
// 	Value int
// 	Left  *BST
// 	Right *BST
// }


// func (tree *BST) FindClosestValue(target int) int {
// 	return tree.findClosestValue(target, tree.Value)
// }
// // Avarage: O(log(n)) time | O(log(n)) space
// // Worst: O(n) time | O(n) space
// func (tree *BST) findClosestValue(target int, closest int) int {
// 	if abcdif(target, closest) > abcdif(target, tree.Value) {
// 		closest = tree.Value
// 	}

// 	if target<tree.Value && tree.Left != nil {
// 		return tree.Left.findClosestValue(target, closest)
// 	}  else if target>tree.Value && tree.Left != nil { 
// 		return tree.Right.findClosestValue(target, closest)
// 	}
// 	return closest
// }

// func abcdif(a, b int) int {
// 	if a>b {
// 		return a-b
// 	}

// 	return b-a
// }







// func main() {
// 	println("dsdsd")
// }


package main

type BST struct {
	Value int
	Left  *BST
	Right *BST
}

// FindClosestValue iteratively with O(n) time and O(1) space
func (tree *BST) FindClosestValue(target int) int {
	closest := tree.Value
	currentNode := tree

	for currentNode != nil {
		if abs(target-closest) > abs(target-currentNode.Value) {
			closest = currentNode.Value
		}

		if target < currentNode.Value {
			currentNode = currentNode.Left
		} else if target > currentNode.Value {
			currentNode = currentNode.Right
		} else {
			break // Exact match found
		}
	}
	return closest
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	// Example: Create BST
	bst := &BST{
		Value: 10,
		Left: &BST{
			Value: 5,
			Left: &BST{
				Value: 2,
			},
			Right: &BST{
				Value: 5,
			},
		},
		Right: &BST{
			Value: 15,
			Right: &BST{
				Value: 22,
			},
		},
	}

	// Find closest value
	target := 12
	println("The closest value to", target, "is", bst.FindClosestValue(target))
}

File: algo_7_bst/src/main.rs
// struct BST {
//     value: i32,
//     left: Option<Box<BST>>,
//     right: Option<Box<BST>>,
// }

// impl BST {
//     // Public method to find the closest value
//     fn find_closest_value(&self, target: i32) -> i32 {
//         self.find_closest_value_helper(target, self.value)
//     }

//     // Avarage: O(log(n)) time | O(log(n)) space
//     // Worst: O(n) time | O(n) space
//     // Recursive helper method
//     fn find_closest_value_helper(&self, target: i32, closest: i32) -> i32 {
//         let current_closest = if (target - closest).abs() > (target - self.value).abs() {
//             self.value
//         } else {
//             closest
//         };

//         if target < self.value {
//             if let Some(left) = &self.left {
//                 return left.find_closest_value_helper(target, current_closest);
//             }
//         } else if target > self.value {
//             if let Some(right) = &self.right {
//                 return right.find_closest_value_helper(target, current_closest);
//             }
//         }
//         current_closest
//     }
// }

// fn main() {
//     // Example: Creating a BST
//     let bst = BST {
//         value: 10,
//         left: Some(Box::new(BST {
//             value: 5,
//             left: Some(Box::new(BST {
//                 value: 2,
//                 left: None,
//                 right: None,
//             })),
//             right: Some(Box::new(BST {
//                 value: 5,
//                 left: None,
//                 right: None,
//             })),
//         })),
//         right: Some(Box::new(BST {
//             value: 15,
//             left: None,
//             right: Some(Box::new(BST {
//                 value: 22,
//                 left: None,
//                 right: None,
//             })),
//         })),
//     };

//     let target = 12;
//     let closest_value = bst.find_closest_value(target);
//     println!("The closest value to {} is {}", target, closest_value);
// }

struct BST {
    value: i32,
    left: Option<Box<BST>>,
    right: Option<Box<BST>>,
}


// FindClosestValue iteratively with O(n) time and O(1) space
impl BST {
    // Find the closest value iteratively
    fn find_closest_value(&self, target: i32) -> i32 {
        let mut closest = self.value;
        let mut current = Some(self);

        while let Some(node) = current {
            if (target - closest).abs() > (target - node.value).abs() {
                closest = node.value;
            }
            if target < node.value {
                current = node.left.as_deref();
            } else if target > node.value {
                current = node.right.as_deref();
            } else {
                break; // Exact match found
            }
        }

        closest
    }
}

fn main() {
    // Example: Create BST
    let bst = BST {
        value: 10,
        left: Some(Box::new(BST {
            value: 5,
            left: Some(Box::new(BST {
                value: 2,
                left: None,
                right: None,
            })),
            right: Some(Box::new(BST {
                value: 5,
                left: None,
                right: None,
            })),
        })),
        right: Some(Box::new(BST {
            value: 15,
            left: None,
            right: Some(Box::new(BST {
                value: 22,
                left: None,
                right: None,
            })),
        })),
    };

    let target = 12;
    let closest_value = bst.find_closest_value(target);
    println!("The closest value to {} is {}", target, closest_value);
}


